src/lib/operators/join_hash/join_hash_steps.hpp

```diff
@@ -442,14 +442,14 @@ void probe(const RadixContainer<RightType>& radix_container,
             } else {
               auto match_found = false;
               for (const auto& row_id : primary_predicate_matching_rows) {
-                if (multi_predicate_join_evaluator->fulfills_all_predicates(row_id, right_row.row_id)) {
+                if (multi_predicate_join_evaluator->satisfies_all_predicates_early_exit(row_id, right_row.row_id)) {
```

Comment:
What's the reason for `_early_exit`?

Reply:
We stop evaluating predicates when one predicate is not satisfied (early exit). The _null_exit variant gives a more detailed evaluation result that is necessary for the AntiDiscardNulls join.
As far as naming is concerned, `satisfies_predicates`and `satisfies_predicates_detailed_result` might be more meaningful.

Reply:
I would assume everybody expects an early exit for "all_predicates" as soon as one does not match.  `satisfies_all_predicates` sound better to me.

And what is a more "detailed evaluation result"?

Reply:
From my understanding the discard version does not need to check the secondary predicates at all when the semi-join value (here, the probe value) is NULL. Am I being wrong here?
As far as I understand the question, the DiscardNULL version is produced by `NOT IN` reformulations and a NULL cannot match here as `NOT IN` implies that the value is checked for equality with all "other" values.

PS: I cannot check the code right now, but please make sure to note that we LEFT relation is the semi relation and the hash join is a special case where we switch the relations. Talking about "only retain values of the right table" might confuse people.